Sources_v3/Controllers/CurrentUserController_Tests.swift

```diff
@@ -0,0 +1,355 @@
+        controller = CurrentUserController(client: client, environment: env.environment)
+        controllerCallbackQueueID = UUID()
+        controller.callbackQueue = .testQueue(withId: controllerCallbackQueueID)
+    }
+    
+    override func tearDown() {
+        weak var weak_env = env
+        weak var weak_client = client
+        weak var weak_controller = controller
+        
+        env = nil
+        client = nil
+        controller = nil
+        controllerCallbackQueueID = nil
+        
+        // We need to assert asynchronously, because there can be some delegate callbacks happening
+        // on the background queue, that keeps the controller alive, until they have finished.
+        AssertAsync {
+            Assert.willBeNil(weak_env)
+            Assert.willBeNil(weak_client)
+            Assert.willBeNil(weak_controller)
+        }
+        
+        super.tearDown()
+    }
+    
+    // MARK: Controller
+
+    func test_initialState() {
+        // Assert client is assigned correctly
+        XCTAssertTrue(controller.client === client)
+        
+        // Assert initial state is correct
+        XCTAssertEqual(controller.state, .inactive)
+        
+        // Assert user is nil
+        XCTAssertNil(controller.currentUser)
+        
+        // Assert unread-count is zero
+        XCTAssertEqual(controller.unreadCount, .noUnread)
+    }
+    
+    func test_startUpdating_changesStateCorrectly_ifCompletesWithAnyError() throws {
+        // Start updating
+        _ = try await(controller.startUpdating)
+        
+        // Assert state changes to `.localDataFetched`
+        XCTAssertEqual(controller.state, .localDataFetched)
+    }
+    
+    func test_startUpdating_stateStaysInactive_ifCompletesWithError() throws {
+        // Update mock observer to throws the error
+        env.currentUserObserverStartUpdatingError = TestError()
+
+        // Start updating
+        _ = try await(controller.startUpdating)
+        
+        // Assert state stays inative
+        XCTAssertEqual(controller.state, .inactive)
+    }
+    
+    func test_startUpdating_propogatesError() throws {
+        // Update mock observer to throws the error
+        env.currentUserObserverStartUpdatingError = TestError()
+        
+        // Start updating and catch the error
+        let startUpdatingError = try await(controller.startUpdating)
+        
+        // Assert error is propogated
+        XCTAssertNotNil(startUpdatingError)
+    }
+    
+    func test_correctDataIsAvailable_whenStartUpdatingCompletes() throws {
+        let unreadCount = UnreadCount(channels: 10, messages: 212)
+        let userPayload: CurrentUserPayload<NoExtraData> = .dummy(userId: .unique, role: .user, unreadCount: unreadCount)
+
+        // Save user to the db
+        _ = try await {
+            client.databaseContainer.write({ session in
+                try session.saveCurrentUser(payload: userPayload)
+            }, completion: $0)
+        }
```

Comment:
This can be `writeSynchronously`

Reply:
fixed in 9b3c74ee2d704845b8cd83448ade208720df8d94